[ENVOY][AUTH] feat: configurable endpoint id passing

commoddity opened the pull request:
Allows configuration of the `External Authorization Server` to pass the Gateway Endpoint ID via either:
- URL Path - eg. `http://path.grove.city/v1/1a2b3c4d`
- Headers - eg. `-H "x-endpoint-id: 1a2b3c4d"`

closes https://github.com/buildwithgrove/path/issues/77

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+type EndpointIDExtractorType string
+
+const (
+	EndpointIDExtractorTypeURLPath EndpointIDExtractorType = "url_path"
```
#PUC with what an example URL looks like

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+
+const (
+	EndpointIDExtractorTypeURLPath EndpointIDExtractorType = "url_path"
+	EndpointIDExtractorTypeHeader  EndpointIDExtractorType = "header"
```
#PUC with what an example curl request looks like

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+	EndpointIDExtractorTypeHeader  EndpointIDExtractorType = "header"
+)
+
+func (e EndpointIDExtractorType) IsValid() bool {
```
```suggestion
// IsValid ensure the endpoint ID extractor type is supported
func (e EndpointIDExtractorType) IsValid() bool {
```

Olshansk commented on local/path/envoy.yaml:
```diff
@@ -0,0 +1,33 @@
+                # Routing configuration for incoming requests.
```
What's with the indent here?

commoddity replied:
File wasn't meant to be committed.

Olshansk commented on envoy/auth_server/main.go:
```diff
@@ -25,15 +25,19 @@ import (
 // TODO_CONSIDER(@commoddity): Make this configurable. See thread here: https://github.com/buildwithgrove/path/pull/52/files/1a3e7a11f159f5b8d3c414f2417f7879bcfab410..258136504608c1269a27047bb9bded1ab4fefcc8#r1859409934
 const port = 10003
 
+// TODO_TECHDEBT(@commoddity): Make these values part of PATH's config YAML and remove the dependency on environment variables.
```
Any chance we can make this an MVP or an upnext?

This doesn't seem like THAT much work.

commoddity replied:
Done.

Olshansk commented on envoy/auth_server/main.go:
```diff
@@ -97,12 +111,23 @@ func main() {
 		panic(err)
 	}
 
+	// Determine which gateway endpoint ID extractor to use
```
Thoughts on creating a small helper for this?

commoddity replied:
Done.

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+	}
+}
+
+// EndpointIDExtractor defines an interface for extracting an endpoint ID
+// from a given source, which could be a URL path or an HTTP header.
```
```suggestion
// EndpointIDExtractor defines an interface for extracting an endpoint ID from a given source.
// This could be a URL path, HTTP header, etc...
```

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+// from a given source, which could be a URL path or an HTTP header.
+type EndpointIDExtractor interface {
+	// Extract extracts the endpoint ID from the provided source.
+	// The sourceType parameter specifies whether the source is a "path" or "header".
```
sourceType no longer exists?

commoddity replied:
It was an old comment; I've updated.

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+// that extracts the endpoint ID from the URL path.
+type URLPathExtractor struct{}
+
+// ExtractGatewayEndpointID extracts the endpoint ID from the URL path.
```
```suggestion
// extractGatewayEndpointID extracts the endpoint ID from the URL path.
```

Olshansk commented on envoy/auth_server/auth/endpoint_id_extractor.go:
```diff
@@ -0,0 +1,81 @@
+// ExtractGatewayEndpointID extracts the endpoint ID from the URL path.
+// The endpoint ID is expected to be the first segment of the path after the pathPrefix (/v1/)
+//
+// eg. http://eth.path.grove.city/v1/1a2b3c4d -> endpointID = "1a2b3c4d"
```
Nice example!

Olshansk commented on local/path/envoy.yaml:
```diff
@@ -0,0 +1,33 @@
+                # Routing configuration for incoming requests.
```
I don't fully understand why we've added this file in this PR?

commoddity replied:
It wasn't meant to be committed; was just there from my local development (there was nothing sensitive in it). I've removed.

Olshansk commented on envoy/auth_server/.env.example:
```diff
@@ -8,3 +8,8 @@ GRPC_HOST_PORT=path_auth_data_server:50051
 # Default is "false" if not set.
```
Do you mind adding a couple Makefile targets to help test a request with the URL & the header?

For reference, something similar to this: https://github.com/pokt-network/poktroll/blob/938f8183b9ab1fd1f6b1bc5d8ff2bba7ab6e2a18/makefiles/relay.mk#L7

commoddity replied:
Added to Makefile and referenced in docs.

Olshansk reviewed (changes requested):
Left comments but forgot to request changes

commoddity commented:
> Left comments but forgot to request changes

Thanks for the review; I've addressed all comments, as well as made updates to the docs to capture the changes in this PR.

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -222,20 +207,105 @@ sequenceDiagram
+If the `ENDPOINT_ID_EXTRACTOR` environment variable is not set, the default `url_path` extractor is used.
+
+:::warning
+Requests are rejected if either of the following are true:
```
OPTIONAL NIT: 

Adding a newline before/after the `:::`

Ditto elsewhere.

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -222,20 +207,105 @@ sequenceDiagram
+:::
+
+:::info
+Regardless of which extractor is used, the Gateway Endpoint ID will always be set in the `x-endpoint-id` header if the reuqest is forwarded to the PATH Service.
```
Is `x-` deprecated? Should we deprecate it?

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -222,20 +207,105 @@ sequenceDiagram
+Regardless of which extractor is used, the Gateway Endpoint ID will always be set in the `x-endpoint-id` header if the reuqest is forwarded to the PATH Service.
+:::
+
+### 4.1 URL Path
```
Try to make titles 
```suggestion
### 4.1 URL Path Endpoint ID Extractor
```

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
````diff
@@ -222,20 +207,105 @@ sequenceDiagram
+  -d '{"jsonrpc": "2.0", "id": 1, "method": "eth_blockNumber" }'
+```
+
+### 4.2 Header
````
```suggestion
### 4.2 Header Endpoint ID Extractor
```

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
````diff
@@ -222,20 +207,105 @@ sequenceDiagram
+```
+
+### 4.2 Header
+
````
Meta point: we might need to find an extension to auto update references to headers.

Olshansk commented on envoy/auth_server/auth/auth_handler.go:
```diff
@@ -20,8 +19,8 @@ import (
+	// Not sure the best way to do this as it is referred to in multiple disparate places (eg. envoy.yaml, PATH's router.go & here)
 	pathPrefix = "/v1/"
 
 	reqHeaderEndpointID          = "x-endpoint-id"    // Set on all service requests
```
Should we add a TODO to remove the `x-`?

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -404,7 +482,9 @@ The custom implementation must use the methods defined in the `GatewayEndpoints`
 
-> 💡 **TIP:** Forking the PADS repo is the easiest way to get started, though any gRPC server implementation that adheres to the `gateway_endpoint.proto` service definition should suffice.
+:::tip
+Forking the PADS repo is the easiest way to get started, though any gRPC server implementation that adheres to the `gateway_endpoint.proto` service definition should suffice.
```
Should they fork it in both the YAML and posters use-cases or just the postgres one?

#PUC

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -222,20 +207,105 @@ sequenceDiagram
+curl http://anvil.localhost:3001/v1 \
+  -X POST \
+  -H "Content-Type: application/json" \
+  -H "x-endpoint-id: endpoint_3" \
```
Wdyt of changing `endpoint_3` to `endpoint_no_auth`.

Ditto regarding the other ones.

Having a semantic name across the board will set a foundation for better docs moving forward.

Olshansk commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -211,30 +212,28 @@ sequenceDiagram
-2. [Header](#222-header)
-
-This is determined by the `ENDPOINT_ID_EXTRACTOR` environment variable in the `auth_server/.env` file.
+1. [URL Path](#221-url-path): e.g. ...
```
@commoddity I thought you'd fill in the examples here

Olshansk commented on makefiles/test_requests.mk:
```diff
@@ -0,0 +1,44 @@
+#######################
```
I ❤️ this file. Ty!

Olshansk reviewed (approved):
@commoddity One minor comment/request, but otherwise g2g!

commoddity reviewed (commented):
Thanks for the review.

commoddity commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -66,7 +68,9 @@ Specifically, this is split into two logical parts:
 
 ### 2.1. Components
```
Actually I changed my mind; I've removed the numbers.

commoddity commented on docusaurus/docs/develop/envoy/introduction.md:
```diff
@@ -222,20 +207,105 @@ sequenceDiagram
+
+The Auth Server may extract the Gateway Endpoint ID from the request in one of two ways:
+
+1. [URL Path](#221-url-path)
```
Nevermind; I've removed the numbers.
